### ISSUE TITLE
[twitter] Broken Python 2 support

### ISSUE THREAD

**Issue by lachs0r** [ytdl-org/youtube-dl#20673]
### Make sure you are using the *latest* version: run `youtube-dl --version` and ensure your version is *2019.04.07*. If it's not, read [this FAQ entry](https://github.com/ytdl-org/youtube-dl/blob/master/README.md#how-do-i-update-youtube-dl) and update. Issues with outdated version will be rejected.
- [x] I've **verified** and **I assure** that I'm running youtube-dl **2019.04.07**

### Before submitting an *issue* make sure you have:
- [x] At least skimmed through the [README](https://github.com/ytdl-org/youtube-dl/blob/master/README.md), **most notably** the [FAQ](https://github.com/ytdl-org/youtube-dl#faq) and [BUGS](https://github.com/ytdl-org/youtube-dl#bugs) sections
- [x] [Searched](https://github.com/ytdl-org/youtube-dl/search?type=Issues) the bugtracker for similar issues including closed ones
- [x] Checked that provided video/audio/playlist URLs (if any) are alive and playable in a browser

### What is the purpose of your *issue*?
- [x] Bug report (encountered problems with youtube-dl)
- [ ] Site support request (request for adding support for a new site)
- [ ] Feature request (request for a new functionality)
- [ ] Question
- [ ] Other

---

### The following sections concretize particular purposed issues, you can erase any section (the contents between triple ---) not applicable to your *issue*

---

### If the purpose of this *issue* is a *bug report*, *site support request* or you are not completely sure provide the full verbose output as follows:

Add the `-v` flag to **your command line** you run youtube-dl with (`youtube-dl -v <your command line>`), copy the **whole** output and insert it here. It should look similar to one below (replace it with **your** log inserted between triple ```):

```
$ youtube-dl -v 'https://twitter.com/Shitty_Future/status/1106486925282684928'
[debug] System config: []
[debug] User config: [u'--netrc', u'--sub-lang', u'deDE', u'--cookies', u'/home/lachs0r/ytdl-cookies.txt', u'--hls-prefer-ffmpeg', u'--fixup', u'warn']
[debug] Custom config: []
[debug] Command-line args: [u'-v', u'https://twitter.com/Shitty_Future/status/1106486925282684928']
[debug] Encodings: locale UTF-8, fs UTF-8, out UTF-8, pref UTF-8
[debug] youtube-dl version 2019.04.07
[debug] Python version 2.7.15 (CPython) - Linux-5.0.5-1-default-x86_64-with-SuSE-20151201-x86_64
[debug] exe versions: ffmpeg 4.1.3, ffprobe 4.1.3
[debug] Proxy map: {}
[twitter] 1106486925282684928: Downloading webpage
[twitter:card] 1106486925282684928: Downloading webpage
Traceback (most recent call last):
  File "/usr/lib64/python2.7/runpy.py", line 174, in _run_module_as_main
    "__main__", fname, loader, pkg_name)
  File "/usr/lib64/python2.7/runpy.py", line 72, in _run_code
    exec code in run_globals
  File "/usr/bin/youtube-dl/__main__.py", line 19, in <module>
  File "/usr/bin/youtube-dl/youtube_dl/__init__.py", line 474, in main

  File "/usr/bin/youtube-dl/youtube_dl/__init__.py", line 464, in _real_main

  File "/usr/bin/youtube-dl/youtube_dl/YoutubeDL.py", line 2008, in download
  File "/usr/bin/youtube-dl/youtube_dl/YoutubeDL.py", line 807, in extract_info
  File "/usr/bin/youtube-dl/youtube_dl/YoutubeDL.py", line 873, in process_ie_result
  File "/usr/bin/youtube-dl/youtube_dl/YoutubeDL.py", line 796, in extract_info
  File "/usr/bin/youtube-dl/youtube_dl/extractor/common.py", line 529, in extract
  File "/usr/bin/youtube-dl/youtube_dl/extractor/twitter.py", line 239, in _real_extract
  File "/usr/bin/youtube-dl/youtube_dl/extractor/common.py", line 2816, in _get_cookies
  File "/usr/lib64/python2.7/Cookie.py", line 583, in __init__
    if input: self.load(input)
  File "/usr/lib64/python2.7/Cookie.py", line 643, in load
    for k, v in rawdata.items():
AttributeError: 'unicode' object has no attribute 'items'
```

### Description of your *issue*, suggested solution and other information

Encountered this issue on openSUSE Tumbleweed.

It seems at least the Twitter extractor no longer works with Python 2, which is unfortunate because the youtube-dl executable uses a shebang that will invoke the Python 2 interpreter on most distributions. I suggest changing the shebang so it uses Python 3 in addition to fixing this issue.

**Comment by lachs0r** [comment]
Of course I omitted the crucial fact that this only breaks when using a cookie file.

**Comment by remitamine** [comment]
> Add the -v flag to your command line you run youtube-dl with (youtube-dl -v <your command line>), copy the whole output and insert it here.

**Comment by lachs0r** [comment]
Updated. Sorry, should have just followed the instructions carefully. Not that it adds anything that wouldn’t have been obvious from the traceback :P

**Comment by remitamine** [comment]
this patch fixes the problem, but not sure if it's the right way to fix it(`str` function normally not called directly in the project):
```diff
diff --git a/youtube_dl/extractor/common.py b/youtube_dl/extractor/common.py
index 0889288f0..efe0006f6 100644
--- a/youtube_dl/extractor/common.py
+++ b/youtube_dl/extractor/common.py
@@ -2813,7 +2813,7 @@ class InfoExtractor(object):
         """ Return a compat_cookies.SimpleCookie with the cookies for the url """
         req = sanitized_Request(url)
         self._downloader.cookiejar.add_cookie_header(req)
-        return compat_cookies.SimpleCookie(req.get_header('Cookie'))
+        return compat_cookies.SimpleCookie(str(req.get_header('Cookie')))
 
     def get_testcases(self, include_onlymatching=False):
         t = getattr(self, '_TEST', None)
```

**Comment by kpmiller** [comment]
I can confirm that I was seeing this same problem, and switching my shell to my python3 install does fix the problem.  (Using the same cookies file.)

**Comment by remitamine** [comment]
`_get_cookies` method no longer called from twitter extractor, superseded by `#23256`.

**Comment by seahorse41** [comment]
I am seeing this same issue after an unattended-upgrade this morning. It was working yesterday. My os is debian.
$ youtube-dl -v -F https://www.dailymotion.com/video/x7rbku3
[debug] System config: []
[debug] User config: [u'--restrict-filenames', u'-o', u'~/Downloads/%(title)s.%(uploader)s.%(id)s.%(format_id)s-%(height)sp.%(ext)s']
[debug] Custom config: []
[debug] Command-line args: [u'-v', u'-F', u'https://www.dailymotion.com/video/x7rbku3']
[debug] Encodings: locale UTF-8, fs UTF-8, out UTF-8, pref UTF-8
[debug] youtube-dl version 2020.01.24
[debug] Python version 2.7.13 (CPython) - Linux-4.9.0-11-686-pae-i686-with-debian-9.11
[debug] exe versions: ffmpeg N-96573-g73aab08d9d, ffprobe N-96573-g73aab08d9d, rtmpdump 2.4
[debug] Proxy map: {}
Traceback (most recent call last):
  File "/usr/lib/python2.7/runpy.py", line 174, in _run_module_as_main
    "__main__", fname, loader, pkg_name)
  File "/usr/lib/python2.7/runpy.py", line 72, in _run_code
    exec code in run_globals
  File "/usr/local/bin/youtube-dl/__main__.py", line 19, in <module>
  File "/usr/local/bin/youtube-dl/youtube_dl/__init__.py", line 474, in main
  File "/usr/local/bin/youtube-dl/youtube_dl/__init__.py", line 464, in _real_main
  File "/usr/local/bin/youtube-dl/youtube_dl/YoutubeDL.py", line 2018, in download
  File "/usr/local/bin/youtube-dl/youtube_dl/YoutubeDL.py", line 796, in extract_info
  File "/usr/local/bin/youtube-dl/youtube_dl/extractor/common.py", line 530, in extract
  File "/usr/local/bin/youtube-dl/youtube_dl/extractor/dailymotion.py", line 235, in _real_extract
  File "/usr/local/bin/youtube-dl/youtube_dl/extractor/dailymotion.py", line 50, in _call_api
  File "/usr/local/bin/youtube-dl/youtube_dl/extractor/dailymotion.py", line 31, in _get_dailymotion_cookies
  File "/usr/local/bin/youtube-dl/youtube_dl/extractor/common.py", line 2831, in _get_cookies
  File "/usr/lib/python2.7/Cookie.py", line 583, in __init__
    if input: self.load(input)
  File "/usr/lib/python2.7/Cookie.py", line 643, in load
    for k, v in rawdata.items():
AttributeError: 'unicode' object has no attribute 'items'

/var/log/unattended-upgrades/unattended-upgrades-dpkg.log says:
Unpacking python-apt-common (1.4.1) over (1.4.0~beta3) ...
Unpacking python3-apt (1.4.1) over (1.4.0~beta3) ...
Unpacking python-apt (1.4.1) over (1.4.0~beta3) ...

$ youtube-dl -U
youtube-dl is up-to-date (2020.01.24)

kpmiller mentioned python3. I don't understand the details of that fix, if it applies here.

**Comment by TaktakTaktouk** [comment]
Any news for Dailymotion.com? we really need a fix for this site. Thank you

**Comment by meaz** [comment]
why is it closed as it doesn't work with dailymotion?
I had to do this:

**Comment by davinkevin** [comment]
Same thing for me, I'll like to have a simpler solution ❤️

**Comment by seahorse41** [comment]
Easy solution is to define an alias, and add it to the home .bash_aliases file
alias youtube-dl='python3 /usr/local/bin/youtube-dl'

**Comment by ldexterldesign** [comment]
I fixed it by upgrading from python 2 to 3.8.2 (macOS, brew and pyenv)

Hope it's useful

Regards

**Comment by tophee** [comment]
Excellent! The latest commit above has fixed the issue. No need for prepending `python3` anymore!
